File: packages/common/src/logger.ts
export enum LogLevel {
  DEBUG = 'DEBUG',
  INFO = 'INFO',
  WARN = 'WARN',
  ERROR = 'ERROR',
  FATAL = 'FATAL'
}

export interface StructuredLog {
  timestamp: string;
  level: LogLevel;
  service: string;
  message: string;
  requestId?: string;
  traceId?: string;
  userId?: string;
  metadata?: Record<string, any>;
  error?: {
    name: string;
    message: string;
    stack?: string;
  };
}

export class Logger {
  constructor(private readonly service: string) {}

  private sanitize(obj: any): any {
    if (!obj || typeof obj !== 'object') return obj;
    const sensitiveKeys = ['password', 'secret', 'totp', 'token', 'privateKey', 'seed', 'apiKey'];
    const cleaned: Record<string, any> = Array.isArray(obj) ? [] : {};

    for (const [key, value] of Object.entries(obj)) {
      if (sensitiveKeys.some(s => key.toLowerCase().includes(s.toLowerCase()))) {
        cleaned[key] = '[REDACTED]';
      } else if (typeof value === 'object') {
        cleaned[key] = this.sanitize(value);
      } else {
        cleaned[key] = value;
      }
    }
    return cleaned;
  }

  private write(level: LogLevel, message: string, meta?: Record<string, any>, err?: Error): void {
    const log: StructuredLog = {
      timestamp: new Date().toISOString(),
      level,
      service: this.service,
      message,
      ...(meta ? { metadata: this.sanitize(meta) } : {}),
      ...(err ? { error: { name: err.name, message: err.message, stack: err.stack } } : {})
    };

    const formatted = JSON.stringify(log);
    if (level === LogLevel.ERROR || level === LogLevel.FATAL) {
      console.error(formatted);
    } else if (level === LogLevel.WARN) {
      console.warn(formatted);
    } else {
      console.log(formatted);
    }
  }

  public debug(message: string, meta?: Record<string, any>): void {
    this.write(LogLevel.DEBUG, message, meta);
  }

  public info(message: string, meta?: Record<string, any>): void {
    this.write(LogLevel.INFO, message, meta);
  }

  public warn(message: string, meta?: Record<string, any>): void {
    this.write(LogLevel.WARN, message, meta);
  }

  public error(message: string, err?: Error, meta?: Record<string, any>): void {
    this.write(LogLevel.ERROR, message, meta, err);
  }

  public fatal(message: string, err?: Error, meta?: Record<string, any>): void {
    this.write(LogLevel.FATAL, message, meta, err);
  }
}

File: syncnode/common/logger.py
import json
import logging
import time
from typing import Any, Dict


SENSITIVE_KEYS = {
    "password",
    "passwordhash",
    "apikey",
    "apisecret",
    "totpsecret",
    "secret",
    "token",
    "authorization",
    "cookie"
}


def sanitize(data: Any) -> Any:
    if isinstance(data, dict):
        cleaned = {}
        for k, v in data.items():
            k_lower = str(k).lower().replace("_", "").replace("-", "")
            if any(s in k_lower for s in SENSITIVE_KEYS) and not any(safe in k_lower for safe in ["status", "count", "type", "tier"]):
                cleaned[k] = "[REDACTED]"
            else:
                cleaned[k] = sanitize(v)
        return cleaned
    elif isinstance(data, list):
        return [sanitize(item) for item in data]
    return data


class Logger:
    def __init__(self, service_name: str = "Syncnode"):
        self.service_name = service_name

    def sanitize(self, data: Any) -> Any:
        return sanitize(data)

    def _log(self, level: str, message: str, meta: Dict[str, Any] = None):
        record = {
            "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            "level": level,
            "service": self.service_name,
            "message": message
        }
        if meta:
            record["meta"] = sanitize(meta)
        print(json.dumps(record))

    def info(self, message: str, meta: Dict[str, Any] = None):
        self._log("INFO", message, meta)

    def warn(self, message: str, meta: Dict[str, Any] = None):
        self._log("WARN", message, meta)

    def error(self, message: str, meta: Dict[str, Any] = None):
        self._log("ERROR", message, meta)

    def debug(self, message: str, meta: Dict[str, Any] = None):
        self._log("DEBUG", message, meta)
